doc/man/isomarkboot.1
.\" This manpage has been automatically generated by docbook2man-spec
.\" from a DocBook document.  docbook2man-spec can be found at:
.\" <http://shell.ipoline.com/~elmert/hacks/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "ISOMARKBOOT" "1" "24 Februar 2004" "isomarkboot" ""
.SH NAME
isomarkboot \- Create bootable CD-ROMs for Linux/Alpha systems.
.SH SYNOPSIS
.sp
\fBisomarkboot\fR \fB/dev/xxx\fR \fBpath/to/bootlx\fR [ \fBpath/to/root.bin\fR ] 
.SH "COPYRIGHT"
.PP
\fBisomarkboot\fR is Copyright (C) 1996 David Mosberger-Tang
.SH "DESCRIPTIONS"
.PP

\fBisomarkboot\fR
enables ISO images created with \fBmkisofs\fR(8) or
retrieved from the net
to be booted on Linux/Alpha systems.
.PP
First you have to prepare your ISO image and mount it loopback on
\fI/dev/xxx\fR (typically \fI/dev/loop0\fR)
(see \fBmount\fR(8) for details). 
.PP
Next you need to locate the aboot-loader (called \fIbootlx\fR)
and the ram disk image containing the root file system (the latter is optional
but usually sensible to specify) which is often called 
\fIroot.bin\fR within your ISO image.
.PP
Finally enable the image to be booted:
.PP
\fBisomarkboot \fI/dev/xxx path/to/bootlx path/to/root.bin\fB\fR
.PP
A typical location of \fIbootlx\fR on the disk image is
\fIboot/bootlx\fR. The path of \fIbootlx\fR
and \fIroot.bin\fR are relative to the root of the image. If
your image is not mounted you have to replace \fI/dev/xxx\fR
with the path to the image.
.PP
Finally you have to burn the CD using e.g.
\fBxcdroast\fR after unmounting the image.
.SH "BUGS"
.PP
\fBisomarkboot\fR is not heavily tested on
architectures other than Linux/Alpha, though it compiles (with
slight patches). Please report bugs if you experience them.
.SH "NOTES"
.PP
\fBisomarkboot\fR is capable of handling symlinks
within the image.
.SH "AUTHOR"
.PP
This man page was written by Helge Kreutzmann <kreutzm@itp.uni-hannover.de> for the Debian GNU/Linux project but may be used by others.
.SH "SEE ALSO"
.PP
\fBaboot\fR(8), \fBmkisofs\fR(8)


ChangeLog
2007-09-11  Helge Kreutzmann <debian@helgefjell.de>
        * Remove audit.h from fs/ext2, unused, and simplify kernel version
	  handling
	* Fix misunderstanding about multi volume ISO images, c.f.
	  http://www.y-adagio.com/public/standards/iso_cdromr/sect_1.htm
	  (used since 2004 for Debian)
	* Support TESTING for isomarkboot on non-alpha architectures
	* Add -Wcast-align and -fno-builtin-printf to main CFLAGS (from Debian)
	* Using kernel headers directly is depreceated and also not necessary.
	  Use only "published" interfaces from linux-libc-dev and import a local
	  copy of the missing asm-alpha/pal.h and asm-alpha/system.h from the 
	  kernel sources for use by head.S and aboot.c.
	  This patch is taken from the Debian version, c.f.
	  http://bugs.debian.org/437296 for details.
	* Adjust headers to make lib/isolib.c portable (tested for i386 on
	  Debian)
        * Allow booting kernels >= 2.6.23rc1. Patch from Richard Henderson
	  via Debian.

2007-09-08  Helge Kreutzmann <debian@helgefjell.de>
        * Minor fixes in German man pages (abootconf.sgml, aboot.sgml)
	* Man pages for isomarkboot belongs in section 1, minor fixes in 
	  German version
	* Add srmbootfat originally from milo, taken from
	  http://www.suse.de/~stepan/source/milo-2.2-18.tar.bz2
	  This has been used by Debian for several years without bug reports
	  already
	* Add netabootwrap to creat bootable images (taken from the Debian
	  package, where it has been used for years already)
	* Add build rules for english man pages
        * Add comments in source for tools/isomarkboot
	* Do not strip binaries by default

2004-04-08  Will Woods <wgwoods@mac.com>
	* Change include path (should fix compilation for 2.6 kernels)
	* Add/fix extra_args patch - allows you to add extra arguments after
	  the configuration number (e.g. "1 console=ttyS0")
	* Fix warning about bcopy
	* Version is now 1.0_pre20040408

2003-12-01  Will Woods <wgwoods@mac.com>
	* Patch for USB console from Jeff.Wiedemeier@hp.com
	* Patch to check kernel memory against console memory descriptors, also
	  from Jeff. These two patches should allow booting on Wildfire.

2003-11-07  Will Woods <wgwoods@mac.com>
	* Changed version number to 1.0pre1 (aboot should hit 1.0 before 
	  they stop making the Alpha, I think)
	* Merged doc changes from 0.10 branch
	* Reverted 0.10 changes out of HEAD (the tag aboot_010pre2 marks the
	  state just before the rollback)

2003-03-21  Will Woods <will.woods@hp.com>
	* Added include/string.h, since we have our own string.c, replaced 
	  #includes of <linux/string.h> with "string.h" since linux/string.h
	  no longer has user-includable prototypes.
	  (thanks to Kelledin <kelledin@skarpsey.dyndns.org> for the idea)
	* Changed #includes of our headers to use quotes instead of <>
	* Don't build with -mcpu=ev4 (default code works on ev4 and is smaller)
	* Cleaned up a couple other warnings 

2003-01-14  Will Woods <will.woods@hp.com>
	* Fixed typos in man pages ("True 64")
	* Bumped version number to 0.9b
	* Released 0.9b. Hooray!

2002-09-18  Will Woods <will.woods@hp.com>
	* branched the 0.9 tree to add bugfixes
	* fixed some compilation errors with newer gcc/glibc versions
	* added newer manpages

2001-11-14  Will Woods <will.woods@compaq.com>
        * sdisklabel/sdisklabel.c: fixed my fix to the cylinder counting bug.
          oops. also added (currently undocumented) "force" flag, to skip
          error checking in sdisklabel.

2001-10-26  Will Woods <will.woods@compaq.com>
        * Released aboot 0.9. Yay!

2001-10-25  Will Woods <will.woods@compaq.com>
	* sdisklabel/sdisklabel.c: fixed cylinder-counting bug

2001-10-11  Will Woods <will.woods@compaq.com>
	* applied a patch from the debian version of aboot. Here's the debian
	maintainer's changelog:

  * New upstream (from CVS)
  * I am calling this version of aboot from CVS 0.8.
  * Minor tweeks to make it compile with kernel 2.4.x headers.
  * e2writeboot now returns non-zero on error.  Closes: #107757
  -- Gregory W. Johnson <gjohnson@physics.clarku.edu>  Wed, 15 Aug 2001

	Since he called his version 0.8, I will call this the 0.9 tree, to 
	avoid confusion. Hence this is aboot 0.9pre1. 

2001-10-09  Will Woods <will.woods@compaq.com>
	* aboot.c, disk.c, include/config.c: replaced ABOOT_MINOR and 
	ABOOT_MAJOR with ABOOT_VERSION. Bumped version number to 0.8pre2.
	* aboot.c: fixed ELF header bug that snuck in somewhere.
	* fs/ext2.c, Makefile: fixes for proper #includes to allow compilation
	of userspace testing aboot against the 2.4 kernel.

2001-10-08  Will Woods <will.woods@compaq.com>
	* fs/ext2.c: Fixed elusive overflow bug. If part of the kernel was 
	past the 524,287th block (given 4096 bytes per block) the byte offset
	was too big to fit in an int. 
	* sdisklabel/sdisklabel.c: borrowed code from fdisk for getting
	the correct size of the disk. 
	
2001-07-10  Will Woods <will.woods@compaq.com>
	* aboot.c, tools/objstrip.c: elf_check_arch takes different arguments
	in Linux 2.4 than it did in 2.2; added an aboot_elf_check_arch which
	does the right thing based on version info from linux/version.h. 
	* disk.c: a function called get_options exists in 2.4 kernels, so
	aboot's get_options function was renamed get_aboot_options(). 
	(both changes based on a patch by Christian Groessler <cpg@aladdin.de>)

2001-04-26  Will Woods <will.woods@compaq.com>
	* disk.c: changed /etc/aboot.conf and aboot.conf back to their
	original settings. If it ain't broke, why fix it?
	* added the man pages in doc/man, written by Helge Kreutzmann
	<kreutzm@itp.uni-hannover.de> (Thanks Helge!)

2001-03-20  Will Woods <will.woods@compaq.com>
	* disk.c: swapped position of /etc/aboot.conf and aboot.conf in the 
	configs array, making /aboot.conf the default.
	* isolib.c: added parse_rock_ridge_inode and get_rock_ridge_symlink
	adapted from rock.c in the linux kernel source. 
	* isolib.c: added iso_follow_link and iso_readdir, adapted from
	fs/ext2.c

tools/isomarkboot.c
/*
 * isomarkboot.c
 *
 * This file is part of aboot, the SRM bootloader for Linux/Alpha
 * Copyright (C) 1996 David Mosberger.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
/*
 * Making an ISO9660 filesystem bootable is straight-forward since all
 * files are contiguous.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <sys/fcntl.h>

#include <config.h>
#include <isolib.h>
#include <iso.h>

const char * prog_name;

static int disk;

void
__memcpy (void * dest, const void * src, size_t n)
{
	memcpy (dest, src, n);
}

long
iso_dev_read (void * buf, long offset, long size)
{
	if (lseek(disk, offset, SEEK_SET) != offset) {
		perror("lseek");
		return -1;
	}
	return read(disk, buf, size);
}

/* Write a 64-bit quantity out into memory in LITTLE ENDIAN order */
static void write_64 (unsigned char* out, unsigned long long in)
{
	out[0] = in & 0xFF;
	out[1] = (in >> 8) & 0xFF;
	out[2] = (in >> 16) & 0xFF;
	out[3] = (in >> 24) & 0xFF;
	out[4] = (in >> 32) & 0xFF;
	out[5] = (in >> 40) & 0xFF;
	out[6] = (in >> 48) & 0xFF;
	out[7] = (in >> 56) & 0xFF;
}

/* Read in a 64-bit LITTLE ENDIAN quantity */
static unsigned long long read_64 (unsigned char *in)
{
	unsigned long long result = 0;

	result |= (unsigned long long) in[0];
	result |= (unsigned long long) in[1] << 8;
	result |= (unsigned long long) in[2] << 16;
	result |= (unsigned long long) in[3] << 24;
	result |= (unsigned long long) in[4] << 32;
	result |= (unsigned long long) in[5] << 40;
	result |= (unsigned long long) in[6] << 48;
	result |= (unsigned long long) in[7] << 56;

	return result;
}

int
main (int argc, char ** argv)
{
	u_int64_t sector[512 / 8], sum;
        struct iso_primary_descriptor vol_desc;
	size_t nbytes, aboot_size;
	off_t aboot_pos;
	int i, aboot_fd;
        int rootbin_fd;
        off_t rootbin_pos;
        char root_start[100];

	prog_name = argv[0];

	if (argc < 3 || argc > 4) {
	    fprintf(stderr, "usage: %s filesys path [root.bin]\n", prog_name);
	    exit(1);
	}
	disk = open(argv[1], O_RDWR);
	if (disk < 0) {
		perror(argv[1]);
		exit(1);
	}

	if (iso_read_super (0, 0) < 0) {
		fprintf(stderr, "%s: cannot mount\n", argv[1]);
		exit(1);
	}

	aboot_fd = iso_open(argv[2]);
	if (aboot_fd < 0) {
		fprintf(stderr, "%s: file not found\n", argv[2]);
		exit(1);
	}

	{
		struct stat buf;
		iso_fstat(aboot_fd, &buf);
		aboot_size = buf.st_size;
	}
		
	aboot_pos = iso_map (aboot_fd, 0);

	printf("%s: %s is at offset %ld and is %lu bytes long\n",
	       prog_name, argv[2], aboot_pos, aboot_size);

	if (lseek(disk, 0, SEEK_SET) != 0) {
		perror("lseek");
		return -1;
	}

	nbytes = read(disk, sector, sizeof(sector));
	if (nbytes != sizeof(sector)) {
		if ((long) nbytes < 0) {
			perror("read");
		} else {
			fprintf(stderr, "%s: short read\n", prog_name);
		}
		exit(1);
	}

	strcpy((char *) sector, "Linux/Alpha aboot for ISO filesystem.");
	write_64 ((unsigned char *) &sector[60], aboot_size /  512);/* sector count */
	write_64 ((unsigned char *) &sector[61], aboot_pos /  512); /* starting LBM */
	write_64 ((unsigned char *) &sector[62], 0);                /* flags */

	/* update checksum: */
	sum = 0;
	for (i = 0; i < 63; i++)
		sum += read_64 ((unsigned char *) &sector[i]);

	write_64 ((unsigned char *) &sector[63], sum);

	if (lseek(disk, 0, SEEK_SET) != 0) {
		perror("lseek");
		return -1;
	}

	nbytes = write(disk, sector, sizeof(sector));
	if (nbytes != sizeof(sector)) {
		if ((long) nbytes < 0) {
			perror("write");
		} else {
			fprintf(stderr, "%s: short write\n", prog_name);
		}
		exit(1);
	}

        if (argc < 4)
          return 0;

	rootbin_fd = iso_open(argv[3]);
	if (rootbin_fd < 0) {
		fprintf(stderr, "%s: file not found\n", argv[3]);
		exit(1);
	}

	rootbin_pos = iso_map (rootbin_fd, 0);
        iso_close(rootbin_fd);

	{
		struct stat buf;
		iso_fstat(rootbin_fd, &buf);
		printf("%s: %s is at offset %ld and is %lu bytes long\n",
		       prog_name, argv[3], rootbin_pos, buf.st_size);
	}

        
	if (lseek(disk, 16*2048, SEEK_SET) != 16*2048) {
		perror("lseek");
		return -1;
	}
	nbytes = read(disk, &vol_desc, sizeof(vol_desc));
	if (nbytes != sizeof(vol_desc)) {
          if ((long) nbytes < 0) {
            perror("read");
          } else {
            fprintf(stderr, "%s: short read\n", prog_name);
          }
          exit(1);
	}

        if (strncmp (vol_desc.id, ISO_STANDARD_ID, sizeof vol_desc.id) != 0) {
          fprintf(stderr,"first volume descriptor has not an ISO_STANDARD_ID!!\n");
          exit(1);
        }
        if (isonum_711 (vol_desc.type) != ISO_VD_PRIMARY) {
          fprintf(stderr,"first volume descriptor is not a primary one!!\n");
          exit(1);
        }
        if (rootbin_pos & 2047) {
          fprintf(stderr,"erreur:rootbin_pos=%ld is not at a isoblock boundary\n",rootbin_pos);
            exit(1);
        }
        sprintf(root_start,"ROOT START=%ld        ",rootbin_pos/2048);
        printf("writing %s in application_data of first volume descriptor\n", root_start); 
        memcpy(vol_desc.application_data,root_start,strlen(root_start));
	if (lseek(disk, 16*2048, SEEK_SET) != 16*2048) {
		perror("lseek");
		return -1;
	}
        
	nbytes = write(disk, &vol_desc, sizeof(vol_desc));
	if (nbytes != sizeof(vol_desc)) {
          if ((long) nbytes < 0) {
			perror("write");
		} else {
                  fprintf(stderr, "%s: short write\n", prog_name);
		}
		exit(1);
	}
        
	return 0;
}
